src/treemodel/selection.js

```diff
@@ -0,0 +1,261 @@
+/**
+ * @license Copyright (c) 2003-2015, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+CKEDITOR.define( [
+	'treemodel/liverange',
+	'treemodel/attributelist',
+	'emittermixin',
+	'utils',
+	'ckeditorerror'
+], ( LiveRange, AttributeList, EmitterMixin, utils, CKEditorError ) => {
+	/**
+	 * Represents a selection that is made on nodes in {@link treeModel.Document}. Selection instance is
+	 * created by {@link treeModel.Document}. In most scenarios you should not need to create an instance of Selection.
+	 *
+	 * @class treeModel.Selection
+	 */
+	class Selection {
+		/**
+		 * Creates an empty selection.
+		 *
+		 * @constructor
+		 */
+		constructor() {
+			/**
+			 * List of attributes set on current selection.
+			 *
+			 * @private
+			 * @property {treeModel.AttributeList} _attrs
+			 */
+			this._attrs = new AttributeList();
+
+			/**
+			 * Stores all ranges that are selected.
+			 *
+			 * @private
+			 * @property {Array.<LiveRange>}
+			 */
+			this._ranges = [];
+
+			/**
+			 * Specifies whether the last added range was added as a backward or forward range.
+			 *
+			 * @private
+			 * @property {Boolean}
+			 */
+			this._lastRangeBackward = false;
```

Comment:
Quoting @scofalik: in Firefox anchor and focus are set based on the last range. Since we don't want to cache these in a form of properties, we keep the last range's direction.

It makes sense that anchor is also taken from the last range when you make multi-range selection using `Cmd/Ctrl`. However, I've been wondering whether this will be also true in tables where you do a single drag to select a column. Turns out it is... I find it a bit odd, but it's legit.

I'm worried though that in case of some changes in the native selection (and such had been discussed at W3C meeting) this won't be true any more. If anchor and focus will be decoupled from the last range what will we need to change in the API and implementation to make it work? Also, what if beside anchor and focus another "point" will be added to the selection API (this actually isn't just a theory, because I remember Ryosuke Niwa mentioning such a possibility).